genrl/deep/agents/dqn/base.py

```diff
@@ -0,0 +1,246 @@
+        self.optimizer = opt.Adam(self.model.parameters(), lr=self.lr_value)
+
+    def update_target_model(self) -> None:
+        """Function to update the target Q model
+
+        Updates the target model with the training model's weights when called
+        """
+        self.target_model.load_state_dict(self.model.state_dict())
+
+    def update_params_before_select_action(self, timestep: int) -> None:
+        """Update necessary parameters before selecting an action
+
+        This updates the epsilon (exploration rate) of the agent every timestep
+
+        Args:
+            timestep (int): Timestep of training
+        """
+        self.timestep = timestep
+        self.epsilon = self.calculate_epsilon_by_frame()
+        self.logs["epsilon"].append(self.epsilon)
+
+    def select_action(
+        self, state: np.ndarray, deterministic: bool = False
+    ) -> np.ndarray:
+        """Select action given state
+
+        Epsilon-greedy action-selection
+
+        Args:
+            state (:obj:`np.ndarray`): Current state of the environment
+            deterministic (bool): Should the policy be deterministic or stochastic
+
+        Returns:
+            action (:obj:`np.ndarray`): Action taken by the agent
+        """
+        if not deterministic:
+            if np.random.rand() < self.epsilon:
+                return np.asarray(self.env.sample())
+
+        state = torch.FloatTensor(state)
+        q_values = self.model(state).detach().numpy()
+        return np.argmax(q_values, axis=-1)
+
+    def get_q_values(self, states: torch.Tensor, actions: torch.Tensor) -> torch.Tensor:
+        """Get Q values corresponding to specific states and actions
+
+        Args:
+            states (:obj:`torch.Tensor`): States for which Q-values need to be found
+            actions (:obj:`torch.Tensor`): Actions taken at respective states
+
+        Returns:
+            q_values (:obj:`torch.Tensor`): Q values for the given states and actions
+        """
+        q_values = self.model(states).gather(1, actions)
+        return q_values
+
+    def get_target_q_values(
+        self, next_states: torch.Tensor, rewards: List[float], dones: List[bool]
+    ) -> torch.Tensor:
+        """Get target Q values for the DQN
+
+        Args:
+            next_states (:obj:`torch.Tensor`): Next states for which target Q-values
+                need to be found
+            rewards (:obj:`list`): Rewards at each timestep for each environment
+            dones (:obj:`list`): Game over status for each environment
+
+        Returns:
+            target_q_values (:obj:`torch.Tensor`): Target Q values for the DQN
+        """
+        next_q_target_values = self.target_model(next_states)
+        max_next_q_target_values = next_q_target_values.max(1)[0]
+
+        target_q_values = rewards + self.gamma * torch.mul(
+            max_next_q_target_values, (1 - dones)
+        )
+        return target_q_values.unsqueeze(-1)
```

Comment:
Can you write down comments before squeeze and unsqueeze operations? 

The reason I'm suggesting this is because it will be easier to debug algorithms so I think its an important addition.

Reply:
Ok I'll do that. In fact, I thought this was one of the main problems with the DQN. Random squeezing, unsqueezing and reshaping everywhere.